
<!doctype html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1">	
<link href="https://fonts.googleapis.com/css?family=Lato:300" rel="stylesheet"> 
<title>PhyloAcc pre-run stats</title>
</head>

<body>
    <div class="row" id="top_grid">
        <div class="col-2-24" id="margin"></div>
        <div class="col-20-24" id="main_header">PhyloAcc pre-run stats</div>
        <div class="col-2-24" id="margin"></div>
    </div>

    <div class="row" id="main-container">
        <div class="row" id="side-nav-row">
            <div class="col-24-24" id="side-nav-container">
                <div class="row" id="nav-header-container">
                    <div class="col-24-24" id="nav-header">Navigation</div>
                </div>
            
                <div class="row" id="nav-content-container">
                    <div class="col-2-24" id="side-nav-margin"></div>
                    <div class="col-22-24" id="nav-links">
                        <a class="nav-link" href="#run-info">1. Run info</a>
                        <a class="nav-link" href="#input-tree">2. Input species tree</a>
                        <a class="nav-link" href="#locus-lens">3. Distribution of locus lengths</a>
                        <a class="nav-link" href="#informative-sites">4. Distribution of informative sites per locus</a>
                        <a class="nav-link" href="#variable-informative-sites">5. Correlation between variable and informative sites</a>
                        <!-- This block is only used when run mode (-r) is adaptive
                        <a class="nav-link" href="#scf">6. Site concordance factors (sCF) per locus</a>
                        <a class="nav-link" href="#scf-tree">7. sCF across all loci</a>
                        <a class="nav-link" href="#bl-scf">8. Branch length vs. sCF for the species tree</a> 
                        -->
                    </div>
                </div>
            </div>
        </div>

        <div class="col-20-24" id="main-content">

            <div class="row content-row" id="run-info-row">
                <div class="col-24-24 content-col" id="run-info">
                    <div class="section-header">1. Run info (phyloacc-out-02-14-2022.09-46-46)</div>
                    <p>The PhyloAcc interface was run on <code>Monday Feb 14, 2022 at 09:46:46 CET</code> on <code>dmp077</code> as follows:</p>
                    
                    <div class="code-focus">
                        <pre>
/usr/users/mko/Tools/PhyloAcc-interface/src/interface/phyloacc_interface.py -a /scratch/users/mko/phyloacc/input/galloseq_gapFixed.fa -b /scratch/users/mko/phyloacc/input/galloseq.part.bed -m /scratch/users/mko/phyloacc/input/nonconserved_4d_named.mod -t HLcinPul1;HLlepAsp1;HLdicExi1;HLlicPen1;HLlicMelCas1;HLphyNov1;HLgraPic1;HLparPun1;HLlorGal1;HLtriMol2;HLaraSol1;HLamaAes1;HLphaSup1;HLfloFus1;HLcalAnn5 -c HLacaPus1;HLapuApu1;HLatrCla1;HLcalPug1;HLchaPel1;HLcliRuf1;HLcolLiv2;HLcorCor3;HLempTra1;HLfalTin1;HLfurRuf1;HLmalCya1;HLmalEle1;HLmelMel1;HLmenNov1;HLnymHol2;HLparMaj1;HLserCan1;HLstrHab1;HLtaeGut4;HLtytAlb2;aptFor1;cucCan1;falChe1;falPer1;ficAlb2;galGal6;halLeu1;opiHoa1;pseHum1 -burnin 1000 -mcmc 7000 -n 1 -p 12 -j 50 -batch 100 -mem 72 -part medium -time 2 -path /usr/product/applsw/phyloacc-1.0/install/bin/PhyloAcc --plot
                        </pre>
                    </div>

                    <div class="line"></div>

                    <div class="sub-header">Alignment and batch info</div>
                    <div class="table-container">
                        <table class="table-content">
                            <tr>
                                <td>Input alignments</td>
                                <td>363476</td>
                            </tr>

                            <tr>
                                <td>Alignments with 0 informative sites (removed)</td>
                                <td>1697</td>
                            </tr>

                            <tr>
                                <td>Alignments for species tree model</td>
                                <td>361779</td>
                            </tr>

                            <tr>
                                <td>Alignments for gene tree model</td>
                                <td>0</td>
                            </tr>

                            <tr>
                                <td>Batches</td>
                                <td>3618</td>
                            </tr>                        

                            <tr>
                                <td>Alignments per batch</td>
                                <td>100</td>
                            </tr>   

                            <tr>
                                <td>Processes to use per batch</td>
                                <td>12</td>
                            </tr>

                            <tr>
                                <td>Batches using species tree model</td>
                                <td>3618</td>
                            </tr>   

                            <tr>
                                <td>Batches using gene tree model</td>
                                <td>0</td>
                            </tr>

                            <tr>
                                <td>Batches to submit to cluster simultaneously</td>
                                <td>50</td>
                            </tr>
                        </table>
                    </div>

                    <p>
                        See the <a href="phyloacc-out-02-14-2022.09-46-46/phyloacc-out-02-14-2022.09-46-46.log">full log file</a> for more info.
                    </p>

                    <p>
                        Below are some summary plots. Raw data is also available in CSV format in the <a href="phyloacc-out-02-14-2022.09-46-46/phyloacc-aln-stats.csv">alignment stats file</a>.
                    </p>

                    <div class="line"></div>
                    <div class="sub-header">Run batches</div>

                    <p>
                        The generated batches can be run with Snakemake with the following command:
                    </p>

                    <div class="code-focus">
                        <pre>
snakemake -p -s /home/mpg08/mko/Nectar/analysis/CNEEanalysis_2021/03_phyloacc/phyloacc-out-02-14-2022.09-46-46/phyloacc-job-files/snakemake/run_phyloacc.smk --configfile /home/mpg08/mko/Nectar/analysis/CNEEanalysis_2021/03_phyloacc/phyloacc-out-02-14-2022.09-46-46/phyloacc-job-files/snakemake/phyloacc-config.yaml --profile /home/mpg08/mko/Nectar/analysis/CNEEanalysis_2021/03_phyloacc/phyloacc-out-02-14-2022.09-46-46/phyloacc-job-files/snakemake/profiles/slurm_profile --cluster-status /home/mpg08/mko/Nectar/analysis/CNEEanalysis_2021/03_phyloacc/slurm_status.py --dryrun
                        </pre>
                    </div>

                    <p>
                        Note that this includes the <code>--dryrun</code> option to check for possible errors before submitting jobs. If everything looks good
                        after running that, feel free to remove the <code>--dryrun</code> option to submit the jobs to your cluster. You may also want to run
                        this in the background using <a href="https://linuxize.com/post/how-to-use-linux-screen/" target="_blank">screen</a>, 
                        <a href="https://linuxize.com/post/getting-started-with-tmux/" target="_blank">tmux</a> or some other terminal multiplexer 
                        to avoid disruptions due to server disconnects.
                    </p>

                    <div class="line"></div>

                    <!-- This block is only displayed when --theta is specified
                    <div class="sub-header">Coalescent tree inference for theta estimation</div>
                    <p>
                        Since some or all of the loci will be processed with the gene tree model and the interface was run with the <code>--theta</code> 
                        option, this Snakemake pipeline includes rules to generate gene trees with<a href="http://www.iqtree.org/" target="_blank">IQ-Tree</a> 
                        and a species tree with branch lengths in coalescent units using <a href="https://github.com/smirarab/ASTRAL" target="_blank">ASTRAL</a>. 
                        The branch lengths from this tree will be used in conjunction with the branch lengths in the input tree (which should be in units 
                        of relative number of substitutions) to estimate the theta parameter used in the gene tree model.
                    </p>
                    -->

                    <!-- This block is only displayed when -l is specified
                    <div class="sub-header">Coalescent tree provided for theta estimation</div>
                    <p>
                        Since some or all of the loci will be processed with the gene tree model, a species tree with branch lengths in coalescent 
                        units was provided: <code>False</code>
                    </p>

                    <p>
                        The branch lengths from this tree will be used in conjunction with the branch lengths in the input tree (which should be in units of 
                        relative number of substitutions) to estimate the theta parameter used in the gene tree model.
                    </p>
                    -->

                </div>
            </div>

            <div class="sep-div"></div>


            <div class="row content-row" id="input-tree-row">
                <div class="col-24-24 content-col" id="input-tree">
                    <div class="section-header">2. Input species tree</div>

                    <div class="table-container">
                        <table class="table-content">
                            <tr>
                                <td>Species</td>
                                <td>45</td>
                            </tr>

                            <tr>
                                <td>Target species</td>
                                <td>15</td>
                            </tr>                        

                            <tr>
                                <td>Conserved species</td>
                                <td>30</td>
                            </tr>   

                            <tr>
                                <td>Outgroup species</td>
                                <td>0</td>
                            </tr>  
                        </table>
                    </div>

                    <div class="row img-row" id="input-tree-container-row">
                        <div class="col-4-24 margin"></div>
                        <div class="col-16-24 img-container" id="input-tree-container">
                            <img id="input-tree-img" src="plots/input-species-tree.png">
                        </div>
                        <div class="col-4-24 margin"></div>
                    </div>

                </div>
            </div>

            <div class="sep-div"></div>

            <div class="row content-row" id="run-info-row">
                <div class="col-24-24 content-col" id="locus-lens">
                    <div class="section-header">3. Distribution of locus lengths</div>

                    <div class="table-container">
                        <table class="table-content">
                            <tr>
                                <td>Average alignment length</td>
                                <td>151.121</td>
                            </tr>

                            <tr>
                                <td>Median alignment length</td>
                                <td>107.0</td>
                            </tr>                        

                            <tr>
                                <td>Average sequence length (no gaps) per alignment</td>
                                <td>35.706</td>
                            </tr>   

                           <tr>
                                <td>Median sequence length (no gaps) per alignment</td>
                                <td>25.133</td>
                            </tr>   

                        </table>
                    </div>

                    <div class="row img-row" id="locus-lens-container-row">
                        <div class="col-2-24 margin"></div>
                        <div class="col-9-24 img-container" id="aln-lens-container">
                            <img id="aln-lens-img" src="plots/aln-len-hist.png">
                        </div>
                        <div class="col-2-24 margin"></div>
                        <div class="col-9-24 img-container" id="seq-lens-container">
                            <img id="seq-lens-img" src="plots/avg-seq-len-nogap-hist.png">
                        </div>
                        <div class="col-2-24 margin"></div>
                    </div>

                </div>
            </div>

            <div class="sep-div"></div>

            <div class="row content-row" id="run-info-row">
                <div class="col-24-24 content-col" id="informative-sites">
                    <div class="section-header">4. Distribution of informative sites per locus</div>

                    <div class="row img-row" id="informative-sites-container-row">
                        <div class="col-2-24 margin"></div>
                        <div class="col-9-24 img-container" id="informative-sites-container">
                            <img id="informative-sites-img" src="plots/informative-sites-hist.png">
                        </div>
                        <div class="col-2-24 margin"></div>
                        <div class="col-9-24 img-container" id="proportion-informative-sites-container">
                            <img id="proportion-informative-sites-img" src="plots/informative-sites-frac-hist.png">
                        </div>
                        <div class="col-2-24 margin"></div>
                    </div>

                </div>
            </div>

            <div class="sep-div"></div>

            <div class="row content-row" id="variable-informative-sites-row">
                <div class="col-24-24 content-col" id="variable-informative-sites">
                    <div class="section-header">5. Correlation between variable and informative sites</div>

                    <div class="row img-row" id="variable-informative-sites-container-row">
                        <div class="col-6-24 margin"></div>
                        <div class="col-12-24 img-container" id="variable-informative-sites-container">
                            <img id="variable-informative-sites-img" src="plots/variable-informative-sites.png">
                        </div>
                        <div class="col-6-24 margin"></div>
                    </div>

                </div>
            </div>

            <div class="sep-div"></div>

            <!-- This block is only used when run mode (-r) is adaptive
            <div class="row content-row" id="scf-row">
                <div class="col-24-24 content-col" id="scf">
                    <div class="section-header">6. sCF distributions per locus</div>

                    <div class="row content-row" id="run-info-row">
                        <div class="col-24-24 content-col" id="scf-info">
                            <p>
                                Site concordance factors (sCF) summarize the number of sites in an alignment that support a branch in a given phylogeny.
                            </p>
                            
                            <p>
                                Briefly, they are computed by sampling quartets of species around each branch in an unrooted tree and counting the number of
                                decisive sites that match the topology of the sampled quartet in the phylogeny. See 
                                <a href="https://doi.org/10.1093/molbev/msaa106" target="_blank">Minh et al. 2020</a> for more information.
                            </p>
                                
                            <p>
                                Here, rather than averaging over branches for all alignments, we calculate sCF for each branch in each alignment separately.
                            </p>
                        </div>
                    </div>

                    <div class="row img-row" id="scf-container-row">
                        <div class="col-2-24 margin"></div>
                        <div class="col-9-24 img-container" id="avg-scf-container">
                            <img id="avg-scf-img" src="plots/avg-scf-per-locus.png">
                        </div>
                        <div class="col-2-24 margin"></div>
                        <div class="col-9-24 img-container" id="low-scf-container">
                            <img id="low-scf-img" src="plots/perc-low-scf-branchers-per-locus.png">
                        </div>
                        <div class="col-2-24 margin"></div>
                    </div>

                </div>
            </div>

            <div class="sep-div"></div>

            <div class="row content-row" id="scf-tree-row">
                <div class="col-24-24 content-col" id="scf-tree">
                    <div class="section-header">7. sCF across all loci</div>
                    <div class="row content-row" id="run-info-row">
                        <div class="col-24-24 content-col" id="scf-info">
                            <p>
                                Here, the species tree is shown with branches labeled by sCF across all branches (as described in
                                <a href="https://doi.org/10.1093/molbev/msaa106" target="_blank">Minh et al. 2020</a>).
                            </p>

                            <p>
                                The tree in Newick format is located <a href="phyloacc-scf.tree">here</a>.
                            </p>
                            
                            <p>
                                The sCF stats per branch can be found in CSV format <a href="phyloacc-scf-stats.csv">here</a>.
                            </p>
                        </div>
                    </div>                            

                    <div class="row img-row" id="scf-tree-container-row">
                        <div class="col-4-24 margin"></div>
                        <div class="col-16-24 img-container" id="scf-tree-container">
                            <img id="scf-tree-img" src="plots/scf-species-tree.png">
                        </div>
                        <div class="col-4-24 margin"></div>
                    </div>

                </div>
            </div>

            <div class="sep-div"></div>

            <div class="row content-row" id="bl-scf-row">
                <div class="col-24-24 content-col" id="bl-scf">
                    <div class="section-header">8. Branch length vs. sCF for the species tree</div>

                    <div class="row img-row" id="bl-scf-container-row">
                        <div class="col-6-24 margin"></div>
                        <div class="col-12-24 img-container" id="bl-scf-container">
                            <img id="bl-scf-img" src="plots/bl-scf.png">
                        </div>
                        <div class="col-6-24 margin"></div>
                    </div>

                </div>
            </div>

            <div class="sep-div"></div>

            -->

        </div>

    </div>


    <div class="row" id="footer">
    <div class="col-1">
        <div id="footer_text">
            <center>Page generated by the <a href="https://github.com/harvardinformatics/PhyloAcc-interface" target="_blank">PhyloAcc interface</a> | Page built: 02/14/2022 10:16:25 CET</center>
        </div>
    </div>

</body>

<style type="text/css" media="screen">
    /*------------------------------------------------------*/
    /* Global page styles */
    body {
        color:#1a1a1a;
        font-family: 'Helvetica', sans-serif;
        font-size: 1.1em;
        background-color: #ececec;
        overflow-y: scroll;
        margin: 0;
    }
    a {
        color:#006ddb;
    }

    /*------------------------------------------------------*/
    /* Grid styles */

    .row { 
        width:100%;
        display: flex;
        justify-content: center;
        align-items: center;
    }

    .col-1-24 { width:4.166667%; display:inline-block; }
    .col-2-24 { width:8.333333%; display:inline-block; }
    .col-3-24 { width:12.5%; display:inline-block; }
    .col-4-24 { width:16.666667%; display:inline-block; }
    .col-5-24 { width:20.833333%; display:inline-block; }
    .col-6-24 { width:25%; display:inline-block; }
    .col-7-24 { width:29.166667%; display:inline-block; }
    .col-8-24 { width:33.333333%; display:inline-block; }
    .col-9-24 { width:37.5%; display:inline-block; }
    .col-10-24 { width:41.666667%; display:inline-block; }
    .col-11-24 { width:45.833333%; display:inline-block; }
    .col-12-24 { width:50%; display:inline-block; }
    .col-13-24 { width:54.166667%; display:inline-block; }
    .col-14-24 { width:58.333333%; display:inline-block; }
    .col-15-24 { width:62.5%; display:inline-block; }
    .col-16-24 { width:66.666667%; display:inline-block; }
    .col-17-24 { width:70.833333%; display:inline-block; }
    .col-18-24 { width:75%; display:inline-block; }
    .col-19-24 { width:79.1666667%; display:inline-block; }
    .col-20-24 { width:83.333333%; display:inline-block; }
    .col-21-24 { width:87.5%; display:inline-block; }
    .col-22-24 { width:91.666667%; display:inline-block; }
    .col-23-24 { width:95.833333%; display:inline-block; }
    .col-24-24 { width:100%; display:inline-block; }
    /* 24 */

    .col-1-12 { width:8.333333%; display:inline-block; }
    .col-2-12 { width:16.666667%; display:inline-block; }
    .col-3-12 { width:25%; display:inline-block; }
    .col-4-12 { width:33.333333%; display:inline-block; }
    .col-5-12 { width:41.666667%; display:inline-block; }
    .col-6-12 { width:50%; display:inline-block; }
    .col-7-12 { width:58.333333%; display:inline-block; }
    .col-8-12 { width:66.666667%; display:inline-block; }
    .col-9-12 { width:75%; display:inline-block; }
    .col-10-12 { width:83.333333%; display:inline-block; }
    .col-11-12 { width:91.666667%; display:inline-block; }
    .col-12-12 { width:100%; display:inline-block; }
    /* 12 */

    .col-1-8 { width:12.5%; display:inline-block; }
    .col-2-8 { width:25%; display:inline-block; }
    .col-3-8 { width:37.5%; display:inline-block; }
    .col-4-8 { width:50%; display:inline-block; }
    .col-5-8 { width:62.5%; display:inline-block; }
    .col-6-8 { width:75%; display:inline-block; }
    .col-7-8 { width:87.5%; display:inline-block; }
    .col-8-8 { width:100%; display:inline-block; }
    /* 8 */

    .col-1-6 { width:16.666667%; display:inline-block; }
    .col-2-6 { width:33.333333%; display:inline-block; }
    .col-3-6 { width:50%; display:inline-block; }
    .col-4-6 { width:66.666667%; display:inline-block; }
    .col-5-6 { width:83.333333%; display:inline-block; }
    .col-6-6 { width:100%; display:inline-block; }
    /* 6 */

    .col-1-4 { width:25%; display:inline-block; }
    .col-2-4 { width:50%; display:inline-block; }
    .col-3-4 { width:75%; display:inline-block; }
    .col-4-4 { width:100%; display:inline-block; }
    /* 4 */

    .col-1-3 { width:33.333333%; display:inline-block; }
    .col-2-3 { width:66.666667%; display:inline-block; }
    .col-3-3 { width:100%; display:inline-block; }
    /* 3 */

    .col-1-2 { width:50%; display:inline-block; }
    .col-2-2 { width:100%; display:inline-block; }
    /* 2 */

    .col-1-1 { width:100%; display:inline-block; }
    /* 1 */

    /*------------------------------------------------------*/
    /* Banner styles */
    #top_grid {
        height:15vh;

        /* background-image: url(html/img/bg1.png); */
        /* https://pixabay.com/illustrations/color-triangle-geometric-textured-2174049/ */

        background: linear-gradient(25deg, rgba(244,251,193,1) 0%, rgba(150,211,196,1) 50%, rgba(50,119,174,1) 100%);
        /* https://cssgradient.io/ */
        
        background-repeat: no-repeat;
        background-clip: border-box;
        background-origin: padding-box;
        -moz-background-size: cover;
        background-size: cover;
    }
    #main_header {
        color:#999999;
        -webkit-text-stroke-width: 0.5px;
        -webkit-text-stroke-color: black;
        font-size:3em;
        padding-top:10px;
        padding-bottom:10px;
        text-align: left;
        vertical-align: middle;
    }

    /*------------------------------------------------------*/
    /* Header styles */

    .section-header {
        background-color: #749abe;
        color: #fff;
        padding: 15px;
        font-size:2em;
        font-weight:bold;
        position: sticky;
        top: 0px;
        /* border-bottom: 2px solid #294157;
        border-left: 2px solid #294157; */
    }
    @media (max-width: 1025px) {
        .section-header  { 		
            text-align:center; 
        }
    }
    .sub-header {
        /* background-color: #749abe; */
        /* color: #fff; */
        padding: 15px;
        font-size:1.5em;
        font-weight:bold;
    }
    .sub-header-2 {
        /* background-color: #749abe; */
        padding: 30px;
        font-size:1.25em;
        font-weight:bold;
    }

    /*------------------------------------------------------*/
    /* Main container styles */

    #main-container {
        display: flex;
        align-items: stretch;
        height: auto;
        min-height: 85vh;
    }

    /*------------------------------------------------------*/
    /* Side nav bar styles */

    #side-nav-row {
        height:97vh;
        background-color: #ffffff;
        border-right:1px solid black;
        position: sticky;
        top: 0px;
    }
    #side-nav-container {
        height:100%;
    }
    #nav-header {
        padding: 15px 0px;
        text-indent: 15px;
        font-size: 2em;
        overflow-wrap: break-word;
        background-color: #294157;
        color: #ffffff;
    }
    .nav-link {
        display: block;
        padding: 10px 10px 10px 0px;
        font-size: 1em;
        text-decoration: none;
        transition: 0.1s linear;

    }
    .nav-link:hover {
        color: #ececec;
        background-color: #56b4e9;
    }
    @media (max-width: 1025px) {
        #side-nav { height:94vh; }
    }

    /*------------------------------------------------------*/
    /* Content styles */

    .content-row p {
        padding: 0 40px;
    }
    .list-container {
        padding-left: 40px;
    }
    .code-focus {
        display: flex;
        align-items: center;
        justify-content: center;
        font-family:'Courier New', Courier, monospace;
        font-size: 1.2em;
    }
    .code-focus pre {
        width: 75%;
        overflow: auto;
        background-color:#ffffff;
    }
    .img-container {
        display: flex;
        align-items: center;
        justify-content: center;
    }
    .img-container img {
        display: inline-block;
        width: 100%;
        height: auto;
    }

    /*------------------------------------------------------*/
    /* Table styles */
    .table-container {
        padding:20px 50px 20px 50px;
        overflow-y:auto;
        display:flex;
        align-items: center;
        justify-content: center;

    }
    .table-content {
        background-color:#ffffff;
        border: 1px solid #333333;
        border-radius: 10px;
        width: 25%;
    }
    .table-content  th {
        background-color:#000000;
        /* background-color: #A9C686; */
        /* background-color: #6B8B39; */
        padding:5px;
        width:12.5%;
        color: #ececec;
        /* border:1px solid red; */
    }
    .table-content  tr:nth-child(even) {
        background-color:#E0DEDE;
    }

    /*------------------------------------------------------*/
    /* Footer grid styles */
    #footer {
        font-family: "Courier New", Courier, monospace;
        background-color: #fff;
        color:#333;
        font-size:0.6em;
        height:3vh;
        display: flex;
        justify-content: center;
        align-content: center;
        flex-direction: column;
        -webkit-box-sizing: border-box;
        -moz-box-sizing: border-box;
        box-sizing: border-box;
        border-top:1px solid #333;
        width:100%;
    }
    @media (max-width: 1025px) {
        #footer { height:6vh; }
        #footer_text { top: 25%; }
    }

    /*------------------------------------------------------*/
    /* Grid spacing styles */
    .sep-div {
        width: 100%;
        height: 2vh;
    }
    #vert_line{
        width:1px;
        height:100%;
        background-color:#d3d3d3;
    }
    .line {
        height:1px;
        background-color:#d3d3d3;
        margin:20px;
    }    
</style>

    